core/dbt/contracts/state.py

```diff
@@ -1,15 +1,21 @@
 from pathlib import Path
 from .graph.manifest import WritableManifest
+
 from .results import RunResultsArtifact
+from .results import FreshnessExecutionResultArtifact
 from typing import Optional
+
 from dbt.exceptions import IncompatibleSchemaException
 
+from dbt.path_utils import FRESHNESS_RESULT_ARCHIVE_DIR_NAME
```

Comment:
@jtcohen6 Ah interesting, we'd considered something like this option before the latest iteration, though that was also hard coded. 

A few additional points to consider.

- How will the user manage previous and current state paths? Wouldn't using `target` for both of those cause dbt to overwrite the state it needs to make the freshness determination per the intended use here?
- What does the availability of these artifacts look like in a dbt cloud deployment job? Do we (theoretically) have access to both the previous and current state, and what are the configurations in which that may not be true?

Reply:
Good questions!

> How will the user manage previous and current state paths? Wouldn't using target for both of those cause dbt to overwrite the state it needs to make the freshness determination per the intended use here?

I believe the user should manage _previous_ state via the same "state" configuration (`--state`/`DBT_ARTIFACT_STATE`) as they do today. I'm advocating for using `target/` as the default path for _current_ state _only_ — i.e. source freshness results, based on the `source freshness` step that just ran. Because `sources.json` is a standalone artifact (unlike `run_results.json`) that is only written by the `source freshness` task, dbt will only overwrite the "current state" (`target/sources.json`) if you run another `dbt source freshness` task. That actually seems right to me! You now have even-more-current freshness results.

> What does the availability of these artifacts look like in a dbt cloud deployment job? Do we (theoretically) have access to both the previous and current state, and what are the configurations in which that may not be true?

Yes! If job deferral is set up, dbt Cloud makes "previous" state accessible by copying artifacts from the deferred-to job and setting `DBT_ARTIFACT_STATE`. Then, "current" state should always be accessible, insofar as the `target/` directory (from previous run steps) is accessible.

Worth confirming / testing to be sure. When this PR is merge-ready, we could test (after merging into `main`) by using the `development (HEAD)` version of dbt Core in dbt Cloud.

Reply:
@matt-winkler I implemented the current state target path `sources.json` logic and verified that custom target paths work!

You can branch off this branch, and we can merge your branch into this branch when your changes are complete.

![image](https://user-images.githubusercontent.com/19176976/154523912-5904c3cc-ab34-4332-9f05-362861a6bc8b.png)
![image](https://user-images.githubusercontent.com/19176976/154524079-3c0696b5-a349-41d0-8fa6-cf03aa6e709f.png)
![image](https://user-images.githubusercontent.com/19176976/154524116-90d8bc46-1524-472c-a750-b6e828a38614.png)

Reply:
FYI: I removed the `CurrentState` class implementation as it would force a refactoring of many other method/class arguments to include it. That's more scope that doesn't fit with this work.